### ISSUE TITLE
Add Dice loss op and Dice loss loss function for semantic segmentation

### ISSUE THREAD

**Issue by AbdelmajidB** [deeplearning4j/deeplearning4j#8645]
Hi everyone,

In semantic segmentation and specially in biomedical segmentation usually we use Dice loss layer as a loss function (presented here:https://arxiv.org/pdf/1606.04797.pdf ), because has a unique advantage that is adaptive to unbalanced data, which is very important for brain tumor segmentation for example because some sub-tumoral regions may only count for a small portion of the whole tumoral volume. the problem is that Dice loss layer  don't exists in DL4J, i tried LossFMeasure with beta=1 with but don't give a good results. 

here is an example with U-net in Python: https://github.com/yihui-he/u-net/blob/master/train.py

Please can someone help me to implement this loss function in DL4J?

Thank you so much.

**Comment by saudet** [comment]
It is possible to implement custom loss layers, yes:
https://deeplearning4j.org/docs/latest/deeplearning4j-nn-custom-layer
https://github.com/eclipse/deeplearning4j-examples/blob/master/dl4j-examples/src/main/java/org/deeplearning4j/examples/misc/lossfunctions/CustomLossExample.java

**Comment by AbdelmajidB** [comment]
Thanks,

But how we can implement that dice loss? I'm a beginner in DL4J.

Please @AlexDBlack have you any idea about this issue??

**Comment by AlexDBlack** [comment]
Right, DL4J support custom losses, but the problem there is to implement the gradients (which aren't easy for something like dice loss, though they are provided in that paper).

An easier alternative may be to implement a SameDiff loss layer - that allows you to avoid implementing the gradients.

Here's an example of implementing mean squared error:
https://github.com/eclipse/deeplearning4j/blob/master/deeplearning4j/deeplearning4j-core/src/test/java/org/deeplearning4j/nn/layers/samediff/testlayers/SameDiffMSELossLayer.java

You would need to change just the defineLayer method to implement dice loss.
In the paper, `p` is equivalent to `layerInput`, and `g` is equivalent to `labels`
So the method would be something like:
```
SDVariable numerator = layerInput.mul(labels).sum(1,2,3).mul(2);  //Skip dimension 0, i.e., batch dimension, for sum
SDVariable denominator = labels.mul(labels).sum(1,2,3).add(layerInput.mul(layerInput).sum(1,2,3)));
SDVariable loss = numerator.div(denominator)
return loss.rsub(1.0); // return 1.0 - loss -> means 0 is best, large values are worst
```

An alternative implementation is shown here: https://www.jeremyjordan.me/semantic-segmentation/

I'll also mark this as a feature request, we can add it directly into DL4J.

**Comment by AbdelmajidB** [comment]
Dear @AlexDBlack, thank you so much, i really appreciate your effort.

I tested it and it gives really good results, this is my SameDiffDiceLossLayer: https://gist.github.com/AbdelmajidB/2a6a73f074925c412570cc63b915fdcd#file-samediffdicelosslayer-java, i added a epsilon to the  denominator to avoid division by zero, and i used it in U-net as here: https://gist.github.com/AbdelmajidB/ee14ab6fe353488a65553adb07b5c3e2. 
I recommend to add it to DL4j as is the most used loss function in semantic segmentation.

I implemented also other interesting loss function called Tversky Loss: https://gist.github.com/AbdelmajidB/039d34ba41513d733995a142193b5093, preseneted here: https://arxiv.org/pdf/1706.05721.pdf. Please @AlexDBlack can you check my implementation?
its implementation in python is presented here: https://github.com/nabsabraham/focal-tversky-unet/blob/master/losses.py

**Comment by AbdelmajidB** [comment]
There is two things about U-net zoo implementation that should be fixed, as its still different to the original paper:
1- Remove the last conv layer ("conv10"), this layer does not exist in the original paper, as descripted in this issue: https://github.com/eclipse/deeplearning4j/issues/6955
2- The kernel size in layer ("conv9-3") is 1 by 1 not 3 by 3 as i  described in this issue: https://github.com/eclipse/deeplearning4j/issues/8214.

You can see my version here: https://gist.github.com/AbdelmajidB/ee14ab6fe353488a65553adb07b5c3e2, the difference is that i use 1 as output feature map in ("conv 9-3") as i'm working on binary segmentation, in the original paper is 2.

**Comment by AlexDBlack** [comment]
@AbdelmajidB great that you got it working.
As for Tversky loss - I haven't come across it before, but looking at the paper and your implementation, I thikn you've got some slight issues in order of operations - i.e., bracket should be after the mul(alpha0 and .mul(beta), not before.

Otherwise looks OK to me

```
    	SDVariable denominator = layerInput.mul(labels).sum(1,2,3)
    			                 .add(layerInput.mul(labels.rsub(1.0)).sum(1,2,3).mul(alpha))
    			                 .add(layerInput.rsub(1.0).mul(labels).sum(1,2,3).mul(beta))
                                         .add(epsilon);
```

As for the unet implementation, you're referring to 2 closed issues. Are you on the latest version?
Or was there something from those issues that didn't get fixed?

**Comment by AbdelmajidB** [comment]
Yes i'm in the last version, i think those closed issues not fixed correctly.
i just want to report about dice loss funcion, i see in the literature three ways of implmentation:
1- adding smal value epsilon  only to the denominator  like here: https://www.jeremyjordan.me/semantic-segmentation/
2-adding small value epsilon to the denominator and the numerator like in tf implementation: https://tensorlayer.readthedocs.io/en/latest/modules/cost.html#tensorlayer.cost.dice_coe
3- adding a smooth with value of 1 to denominator  and the  numerator like here: https://github.com/nabsabraham/focal-tversky-unet/blob/347d39117c24540400dfe80d106d2fb06d2b99e1/losses.py#L8

**Comment by agibsonccc** [comment]
Closing — this issue is over 5 years old with no recent activity. The codebase has been significantly rewritten. If this is still relevant, please open a new issue.
